File: abc/abc150/abc150_c.cs
// https://atcoder.jp/contests/abc150/submissions/9396735

using System;
using System.Collections.Generic;

namespace CSharp
{
    class Program
    {
        static void Main()
        {
            var N = int.Parse(Console.ReadLine());
            var P = Console.ReadLine().Replace(" ", "");
            var Q = Console.ReadLine().Replace(" ", "");

            var s = string.Empty;
            for (int i = 1; i < N + 1; i++)
            {
                s += i.ToString();
            }

            var l = Permutation(s);

            var a = 0;
            var b = 0;
            for (int i = 0; i < l.Count; i++)
            {
                if (l[i] == P)
                {
                    a = i;
                }
                if (l[i] == Q)
                {
                    b = i;
                }
            }
            Console.WriteLine(Math.Abs(a - b));
        }

        static List<string> Permutation(string s)
        {
            if (s.Length == 1)
            {
                return new List<string>() { s };
            }
            var result = new List<string>();
            for (int i = 0; i < s.Length; i++)
            {
                foreach (var t in Permutation(s.Remove(i, 1)))
                {
                    result.Add(s[i] + t);
                }
            }
            return result;
        }
    }
}

File: abc/abc150/abc150_c.fs
// https://atcoder.jp/contests/abc150/submissions/12184961

let N = stdin.ReadLine() |> int
let P = stdin.ReadLine().Replace(" ", "")
let Q = stdin.ReadLine().Replace(" ", "")

let rec distribution e =
    function
    | [] -> [ [ e ] ]
    | x :: xs' as xs ->
        (e :: xs) :: [ for xs in distribution e xs' -> x :: xs ]

let rec permutation =
    function
    | [] -> [ [] ]
    | e :: xs -> List.collect (distribution e) (permutation xs)

let find lst s =
    lst
    |> List.find (fun (x, _) -> x = s)
    |> snd

permutation [ 1..N ]
|> List.sort
|> List.map (fun lst -> System.String.Concat(Array.ofList lst))
|> List.mapi (fun i x -> x, (i + 1))
|> fun lst -> abs (find lst P - find lst Q)
|> stdout.WriteLine
